Neos.Neos/Classes/Fusion/Cache/NeosFusionContextSerializer.php

```diff
@@ -0,0 +1,130 @@
+
+/**
+ * Serializer for Fusion's [at]cache.context values
+ *
+ * Implements special handing for serializing {@see Node} objects in fusions cache context:
+ *
+ * ```
+ * [at]cache {
+ *   mode = 'uncached'
+ *   context {
+ *     1 = 'node'
+ *   }
+ * }
+ * ```
+ *
+ * The property mapper cannot be relied upon to serialize nodes, as this is willingly not implemented.
+ *
+ * Serializing falls back to Fusion's standard {@see FusionContextSerializer} which uses Flow's property mapper.
+ *
+ * @internal
+ */
+final class NeosFusionContextSerializer implements NormalizerInterface, DenormalizerInterface
+{
+    public function __construct(
+        private readonly FusionContextSerializer $fusionContextSerializer,
+        private readonly ContentRepositoryRegistry $contentRepositoryRegistry
+    ) {
+    }
+
+    /**
+     * @param array<int|string,mixed> $context
+     */
+    public function denormalize(mixed $data, string $type, string $format = null, array $context = [])
+    {
+        if ($type === Node::class) {
+            return $this->deserializeNode($data);
+        }
+        return $this->fusionContextSerializer->denormalize($data, $type, $format, $context);
+    }
+
+    /**
+     * @param array<int|string,mixed> $context
+     * @return array<int|string,mixed>
+     */
+    public function normalize(mixed $object, string $format = null, array $context = [])
+    {
+        if ($object instanceof Node) {
+            return $this->serializeNode($object);
+        }
+        return $this->fusionContextSerializer->normalize($object, $format, $context);
+    }
+
+    /**
+     * @param array<int|string,mixed> $serializedNode
+     */
+    private function deserializeNode(array $serializedNode): Node
+    {
+        $contentRepositoryId = ContentRepositoryId::fromString($serializedNode['contentRepositoryId']);
+
+        $contentRepository = $this->contentRepositoryRegistry->get($contentRepositoryId);
+
+        $workspace = $contentRepository->getWorkspaceFinder()->findOneByName(WorkspaceName::fromString($serializedNode['workspaceName']));
+        if (!$workspace) {
+            throw new \RuntimeException(sprintf('Could not find workspace while trying to convert node from array %s.', json_encode($serializedNode)), 1699782153);
+        }
+
+        $subgraph = $contentRepository->getContentGraph()->getSubgraph(
+            $workspace->currentContentStreamId,
+            DimensionSpacePoint::fromArray($serializedNode['dimensionSpacePoint']),
+            $workspace->isPublicWorkspace()
+                ? VisibilityConstraints::frontend()
+                : VisibilityConstraints::withoutRestrictions()
+        );
+
+        $node = $subgraph->findNodeById(NodeAggregateId::fromString($serializedNode['nodeAggregateId']));
+        if (!$node) {
+            throw new \RuntimeException(sprintf('Could not find serialized node %s.', json_encode($serializedNode)), 1699782153);
```

Comment:
Hm, isn't this a probable case that a node was stored which doesn't exist anymore now?
Should we fail hard here, or do we have another way of handling this?

Reply:
That's not the same, in 8.3 this Error would result in the PropertyMapper returning null, thus mostly just breaking rejndering of the sub part here that needs the node. I think that's saner. You cannot control editor behavior...

Reply:
thanks for the hint ^^ should we log anything in the silent fail case?

Reply:
Aaaand adjusted to return null ;)  If the workspace doesnt exist, i will still throw as this is most likely a fundamental error and not an editor hickup... i think ...